Add setting use brackets as a namespace separator

alxfv opened the pull request:
## Add a new setting that allows using square brackets as a namespace separator. 

In a company I work for we use square brackets as a struct separator.

- [*] Tests exist or have been written that cover this particular change.

@go-playground/admins

deankarn commented:
@alxfv I will have to review carefully, there is a good chance this will conflict with the map and/or array decoding logic which uses square brackets as its separators.

alxfv commented:
> @alxfv I will have to review carefully, there is a good chance this will conflict with the map and/or array decoding logic which uses square brackets as its separators.

Thanks! I will add more test to try to find possible conflicts.

alxfv commented:
Hello again!

Improved some tests: 
1. Use [table tests](https://github.com/golang/go/wiki/TableDrivenTests) to run tests with and without `useBrackest` setting. 
2. Added in `TestDecoderStruct` all cases from `TestDecoderMapKeys` and haven't found any issues with decoding structs and map with square brackets in the same values.
3. Added `TestTestDecoderStructMapConflict` that shows what happens if map key and struct field have the same value:
https://github.com/go-playground/form/blob/ad71d565a0abedf3072d7a19e7a2ff00ebe18bdf/decoder_test.go#L537-L585

`"Phone[Number]": []string{"111"},` with `usingBrackets = true` will be decoded to struct or/and a map and it looks as expected.

Also wanted to tell how this pull request implements this feature. It changes only `namespace` slice here:

https://github.com/go-playground/form/blob/91efee7b2a12a43398c6cd5318c7e435e1ca4e6e/decoder.go#L162-L169

And later on in `findAlias` we will get `Phone[Number]` instead of `Phone.Number`, that's all. 
https://github.com/go-playground/form/blob/91efee7b2a12a43398c6cd5318c7e435e1ca4e6e/decoder.go#L34-L41

deankarn commented:
I think that this will likely be ok then. 

I only have one ask to make this more generic for any future requests can the PR be modified to use a struct namespace `prefix` and struct namespace `suffix`?

This will help keep the library flexible to future use cases and remove the `if` logic associated with useBarackets :)

alxfv commented:
> I think that this will likely be ok then.
> 
> I only have one ask to make this more generic for any future requests can the PR be modified to use a struct namespace `prefix` and struct namespace `suffix`?
> 
> This will help keep the library flexible to future use cases and remove the `if` logic associated with useBarackets :)

Great idea, I will do it today.
